File: src/test/csharp/eu/sig/training/ch03/binarytree/BinaryTreeSearchTest.cs
using System;
using NUnit.Framework;

namespace eu.sig.training.ch03.binarytree
{
    [TestFixture]
    public class BinaryTreeSearchTest
    {

        [Test]
        public void testSimpleTree()
        {
            BinaryTreeNode<Int32> root = new BinaryTreeNode<Int32>(5);
            root.Insert(3);
            root.Insert(10);
            root.Insert(12);

            Assert.AreEqual(eu.sig.training.ch03.binarytree.v1.BinaryTreeSearch.CalculateDepth(root, 12), 2);
            Assert.AreEqual(eu.sig.training.ch03.binarytree.v2.BinaryTreeSearch.CalculateDepth(root, 12), 2);
            Assert.AreEqual(eu.sig.training.ch03.binarytree.v3.BinaryTreeSearch.CalculateDepth(root, 12), 2);
        }

        [Test]
        public void testDepth0()
        {
            BinaryTreeNode<Int32> root = new BinaryTreeNode<Int32>(2);
            root.Insert(5);
            root.Insert(1);
            root.Insert(16);

            Assert.AreEqual(eu.sig.training.ch03.binarytree.v1.BinaryTreeSearch.CalculateDepth(root, 2), 0);
            Assert.AreEqual(eu.sig.training.ch03.binarytree.v2.BinaryTreeSearch.CalculateDepth(root, 2), 0);
            Assert.AreEqual(eu.sig.training.ch03.binarytree.v3.BinaryTreeSearch.CalculateDepth(root, 2), 0);
        }

        [Test]
        public void testNotInTree()
        {
            BinaryTreeNode<Int32> root = new BinaryTreeNode<Int32>(2);
            root.Insert(5);
            root.Insert(1);
            root.Insert(16);
            bool exceptionCaught_v1 = false;
            bool exceptionCaught_v2 = false;
            bool exceptionCaught_v3 = false;

#pragma warning disable 168
            try
            {
                Assert.AreEqual(eu.sig.training.ch03.binarytree.v1.BinaryTreeSearch.CalculateDepth(root, 17), 0);
            }
            catch (TreeException e)
            {
                exceptionCaught_v1 = true;
            }
            try
            {
                Assert.AreEqual(eu.sig.training.ch03.binarytree.v2.BinaryTreeSearch
                    .CalculateDepth(root, 17), 0);
            }
            catch (TreeException e)
            {
                exceptionCaught_v2 = true;
            }
            try
            {
                Assert.AreEqual(eu.sig.training.ch03.binarytree.v3.BinaryTreeSearch
                    .CalculateDepth(root, 17), 0);
            }
            catch (TreeException e)
            {
                exceptionCaught_v3 = true;
            }
            Assert.IsTrue(exceptionCaught_v1);
            Assert.IsTrue(exceptionCaught_v2);
            Assert.IsTrue(exceptionCaught_v3);
#pragma warning restore 168
        }
    }
}

File: src/test/java/eu/sig/training/ch03/binarytree/BinaryTreeSearchTest.java
package eu.sig.training.ch03.binarytree;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

public class BinaryTreeSearchTest {

    @Test
    public void testSimpleTree() {
        BinaryTreeNode<Integer> root = new BinaryTreeNode<Integer>(5);
        root.insert(3);
        root.insert(10);
        root.insert(12);

        assertEquals(eu.sig.training.ch03.binarytree.v1.BinaryTreeSearch.calculateDepth(root, 12), 2);
        assertEquals(eu.sig.training.ch03.binarytree.v2.BinaryTreeSearch.calculateDepth(root, 12), 2);
        assertEquals(eu.sig.training.ch03.binarytree.v3.BinaryTreeSearch.calculateDepth(root, 12), 2);
    }

    @Test
    public void testDepth0() {
        BinaryTreeNode<Integer> root = new BinaryTreeNode<Integer>(2);
        root.insert(5);
        root.insert(1);
        root.insert(16);

        assertEquals(eu.sig.training.ch03.binarytree.v1.BinaryTreeSearch.calculateDepth(root, 2), 0);
        assertEquals(eu.sig.training.ch03.binarytree.v2.BinaryTreeSearch.calculateDepth(root, 2), 0);
        assertEquals(eu.sig.training.ch03.binarytree.v3.BinaryTreeSearch.calculateDepth(root, 2), 0);
    }

    @Test
    public void testNotInTree() {
        BinaryTreeNode<Integer> root = new BinaryTreeNode<Integer>(2);
        root.insert(5);
        root.insert(1);
        root.insert(16);
        boolean exceptionCaught_v1 = false;
        boolean exceptionCaught_v2 = false;
        boolean exceptionCaught_v3 = false;

        try {
            assertEquals(eu.sig.training.ch03.binarytree.v1.BinaryTreeSearch.calculateDepth(root, 17), 0);
        } catch (TreeException e) {
            exceptionCaught_v1 = true;
        }
        try {
            assertEquals(eu.sig.training.ch03.binarytree.v2.BinaryTreeSearch
                .calculateDepth(root, 17), 0);
        } catch (TreeException e) {
            exceptionCaught_v2 = true;
        }
        try {
            assertEquals(eu.sig.training.ch03.binarytree.v3.BinaryTreeSearch
                .calculateDepth(root, 17), 0);
        } catch (TreeException e) {
            exceptionCaught_v3 = true;
        }
        assertTrue(exceptionCaught_v1);
        assertTrue(exceptionCaught_v2);
        assertTrue(exceptionCaught_v3);
    }

}
